src/momento/internal/synchronous/_vector_index_data_client.py

```diff
@@ -0,0 +1,108 @@
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Optional
+
+from momento_wire_types import vectorindex_pb2 as vectorindex_pb
+from momento_wire_types import vectorindex_pb2_grpc as vectorindex_grpc
+
+from momento import logs
+from momento.auth import CredentialProvider
+from momento.config import VectorIndexConfiguration
+from momento.errors import convert_error
+from momento.internal._utilities import _validate_index_name, _validate_top_k
+from momento.internal.synchronous._vector_index_grpc_manager import (
+    _VectorIndexDataGrpcManager,
+)
+from momento.requests.vector_index.item import Item
+from momento.responses.vector_index import (
+    Search,
+    SearchResponse,
+    UpsertItemBatch,
+    UpsertItemBatchResponse,
+)
+from momento.responses.vector_index.data.search import SearchHit
+
+
+class _VectorIndexDataClient:
+    """Internal vector index data client."""
+
+    def __init__(self, configuration: VectorIndexConfiguration, credential_provider: CredentialProvider):
+        endpoint = credential_provider.cache_endpoint
+        self._logger = logs.logger
+        self._logger.debug("Vector index data client instantiated with endpoint: %s", endpoint)
+        self._endpoint = endpoint
+
+        default_deadline: timedelta = configuration.get_transport_strategy().get_grpc_configuration().get_deadline()
+        self._default_deadline_seconds = int(default_deadline.total_seconds())
+
+        self._grpc_manager = _VectorIndexDataGrpcManager(configuration, credential_provider)
+
+    @property
+    def endpoint(self) -> str:
+        return self._endpoint
+
+    def upsert_item_batch(
+        self,
+        index_name: str,
+        items: list[Item],
+    ) -> UpsertItemBatchResponse:
+        try:
+            self._log_issuing_request("UpsertItemBatch", {"index_name": index_name})
+            _validate_index_name(index_name)
+            request = vectorindex_pb.UpsertItemBatchRequest(
+                index_name=index_name,
+                items=[item.to_proto() for item in items],
+            )
+
+            self._build_stub().UpsertItemBatch(request, timeout=self._default_deadline_seconds)
+
+            self._log_received_response("UpsertItemBatch", {"index_name": index_name})
+            return UpsertItemBatch.Success()
+        except Exception as e:
+            self._log_request_error("set", e)
+            return UpsertItemBatch.Error(convert_error(e))
+
+    def search(
+        self, index_name: str, query_vector: list[float], top_k: int, metadata_fields: Optional[list[str]] = None
```

Comment:
I think it's probably okay to change this in a subsequent PR if we need to, but we might need a wrapper type for this `list[float]` here, right? for forward-compat with numpy etc.?